Refactoring NodeStateManager etc. to support forecasting functionality

kaituo opened the pull request:
### Description
This commit extends the codebase to support both Anomaly Detection (AD) and forecasting. It contains a mixture of refactoring, renaming, removal of unused code, and package moving tasks. Here are the details:

Refactoring:
- `NodeStateManager.getAnomalyDetector` is now `getConfig`, with added functionality to fetch a Forecaster. The method comments are updated for clarity.
- Existing methods (`getFeatureSamplesForPeriods`, `getColdStartSamplesForPeriods`, `createPreviewSearchRequest`, `getMinDataTime`) have been added in `SearchFeatureDao` to handle forecasting logic.
- Adjusted `SecurityClientUtil` and `ParseUtils` to handle forecasting logic.
- Cleaned up `NodeState` to differentiate state for AD and forecasting.

Renaming:
- `AnomalyDetectorJob` is renamed to `Job` to facilitate reuse for forecasting.
- `NodeStateManager.getAnomalyDetectorJob` is renamed to `getJob`.
- Certain settings in `AnomalyDetectorSettings` are renamed to reflect they are meant for the AD setting. They have been marked as deprecated and new settings are used in `TimeSeriesSettings` instead.
- `IndexAnomalyDetectorJobActionHandler.getAnomalyDetectorJobForWrite` is renamed to `getJobForWrite`.
- `ADSafeSecurityInjector` is renamed to `TimeSeriesSafeSecurityInjector`.

Removing unused code:
- Synchronous code in `ClientUtil`, `IndexUtils`, and `CheckpointDao` is removed.
- The unused class `Throttler` is deleted.
- Mapping file names are changed, and the code referencing these files is adjusted.

Package moving:
- Several classes (`ClientUtil`, `MultiResponsesDelegateActionListener`, `SafeSecurityInjector`, `SecurityUtil`, `ExceptionUtil`, `SearchFeatureDao`, `CleanState`, `ExpiringState`, `MaintenanceState`, `NodeState`, `SingleStreamModelIdMapper`, `BackPressureRouting`) are moved to the respective `org.opensearch.timeseries` packages.

Miscellaneous:
- Fixed compiler failures caused by changes in https://github.com/opensearch-project/OpenSearch/pull/8730 by replacing `DoubleArrayList` with `java.util.ArrayList`.
- Updates the Backwards Compatibility (bwc) version to align with the core's incremented bwc version as per [OpenSearch PR #8670](https://github.com/opensearch-project/OpenSearch/pull/8670). This change prevents the issue described in [OpenSearch Issue #5076](https://github.com/opensearch-project/OpenSearch/issues/5076).

Testing:
- Executed a `gradle build`.
- Added new tests for `ClientUtil` and `NodeStateManager`.

### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/866

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

sudiptoguha commented on src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:
```diff
@@ -410,13 +412,13 @@ private void trainModelFromDataSegments(
      */
     private void getEntityColdStartData(String detectorId, Entity entity, ActionListener<Optional<List<double[][]>>> listener) {
-        ActionListener<Optional<AnomalyDetector>> getDetectorListener = ActionListener.wrap(detectorOp -> {
+        ActionListener<Optional<? extends Config>> getDetectorListener = ActionListener.wrap(detectorOp -> {
```
nice.

sudiptoguha commented on src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:
```diff
@@ -123,15 +125,17 @@ private AnomalyDetectorSettings() {}
-    public static final Setting<TimeValue> BACKOFF_MINUTES = Setting
+    // Use TimeSeriesSettings.BACKOFF_MINUTES
+    @Deprecated
+    public static final Setting<TimeValue> AD_BACKOFF_MINUTES = Setting
```
deprecated for next PR or is the idea to keep this because this is in use? Either answer works, just curious.

kaituo replied:
It's latter.

sudiptoguha commented on src/main/java/org/opensearch/ad/util/ClientUtil.java:
```diff
@@ -1,332 +0,0 @@
-/*
```
Maybe add in the PR header if this would have any effects? In general less code is always better.

kaituo replied:
The code in question is not utilized within the current system, thus it can be safely removed without impacting any existing functionalities.

sudiptoguha commented on src/main/java/org/opensearch/timeseries/feature/SearchFeatureDao.java:
```diff
@@ -899,15 +912,16 @@ public void getColdStartSamplesForPeriods(
+    private SearchRequest createColdStartFeatureSearchRequest(Config detector, List<Entry<Long, Long>> ranges, Optional<Entity> entity) {
         try {
-            SearchSourceBuilder searchSourceBuilder = ParseUtils.generateEntityColdStartQuery(detector, ranges, entity, xContent);
+            SearchSourceBuilder searchSourceBuilder = ParseUtils.generateColdStartQuery(detector, ranges, entity, xContent);
```
The query is still specific to entity, right?  ... but the name change is ok.

kaituo replied:
It is not specific to entity anymore as I change entity field to Optional. People can set it to empty if it is not specific to entity.

amitgalitz replied:
Did we have a separate call of this method without entity? Or just making the code more generic for future development?

sudiptoguha commented on src/main/java/org/opensearch/timeseries/function/BiCheckedFunction.java:
```diff
@@ -0,0 +1,11 @@
+package org.opensearch.timeseries.function;
+
+@FunctionalInterface
+public interface BiCheckedFunction<T, F, R, E extends Exception> {
```
Nice.

sudiptoguha commented on src/main/java/org/opensearch/timeseries/util/ParseUtils.java:
```diff
@@ -515,34 +530,50 @@ public static void getDetector(
-    public static void onGetAdResponse(
+    /**
+     * This method processes the GetResponse and applies the consumer function if the user has permissions
+     * or if the filterByBackendRole is disabled. It uses a ConfigFactory to parse the correct type of Config.
```
What is the expected behavior if the filterByBackendRole is disabled? Maybe add the behaviors here?

kaituo replied:
We will not check backend role related user permission if  if the filterByBackendRole is disabled. Yeah, will add.

sudiptoguha commented on src/test/java/org/opensearch/ad/settings/AnomalyDetectorSettingsTests.java:
```diff
@@ -333,8 +342,10 @@ public void testSettingsGetValueWithLegacyFallback() {
+            .put("plugins.timeseries.max_retry_for_unresponsive_node", 10)
             .put("opendistro.anomaly_detection.cooldown_minutes", "11m")
             .put("opendistro.anomaly_detection.backoff_minutes", "12m")
+            .put("plugins.timeseries.backoff_minutes", "12m")
```
question: are these backoffs surfaced to the user? Can be helpful to them in understanding delays etc.

kaituo replied:
yes, they are surfaced to the user and user can adjust them.

sudiptoguha reviewed (approved):
Looks good - lot of files but most changes are simple. One question: what'd be the side effects in terms of lost abilities to an user (if any) of removing the synchronous code? Maybe add that to the PR header?

kaituo commented:
> Looks good - lot of files but most changes are simple. One question: what'd be the side effects in terms of lost abilities to an user (if any) of removing the synchronous code? Maybe add that to the PR header?

The code in question is not utilized within the current system, thus it can be safely removed without impacting any existing functionalities.

kaituo commented:
bwc test failed due to failing to find 2.10 build.

> Can't get https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/2.10.0/latest/linux/x64/tar/builds/opensearch/plugins/opensearch-anomaly-detection-2.10.0.0.zip to /home/runner/work/anomaly-detection/anomaly-detection/src/test/resources/org/opensearch/ad/bwc/anomaly-detection/2.10.0.0/opensearch-anomaly-detection-2.10.0.0.zip

amitgalitz commented on src/main/java/org/opensearch/ad/settings/AnomalyDetectorSettings.java:
```diff
@@ -123,15 +125,17 @@ private AnomalyDetectorSettings() {}
         );
 
-    public static final Setting<TimeValue> COOLDOWN_MINUTES = Setting
+    public static final Setting<TimeValue> AD_COOLDOWN_MINUTES = Setting
```
Is this so we have ability to have a different request timeout between AD and forecasting? Will you add these to the forecaster settings? I don't see a `FORECAST_MAX_RETRY_FOR_UNRESPONSIVE_NODE` or `FORECAST_COOLDOWN_MINUTES`

kaituo replied:
I changed to add a common colldown and max retry setting in TimeSeriesSettings. Please see my changes in src/main/java/org/opensearch/timeseries/TimeSeriesAnalyticsPlugin.java

amitgalitz commented on src/main/java/org/opensearch/timeseries/NodeStateManager.java:
```diff
@@ -235,68 +157,140 @@ public boolean isMuted(String nodeId, String detectorId) {
+                Config config = configParser.apply(parser, response.getId());
+
+                // end execution if all features are disabled
+                if (config.getEnabledFeatureIds().isEmpty()) {
```
Would there be cases when detector is already running here and then we see features are disbaled and then we throw the endRunException? Would the features be disabled in this case cause of some other error? or this is incase features were never enabled in the first place?

kaituo replied:
When detector is running, we cannot edit detector so cannot disable features. Disabling is done before starting a detector.

amitgalitz commented on src/main/java/org/opensearch/timeseries/NodeStateManager.java:
```diff
@@ -235,68 +157,140 @@ public boolean isMuted(String nodeId, String detectorId) {
      */
-    public boolean hasRunningQuery(AnomalyDetector detector) {
-        return clientUtil.hasRunningQuery(detector);
+    public <T> void getConfig(
```
I see we previously didn't have a getDetector here that takes a consumer function as well, what are the main reasons this is added now as an option?

kaituo replied:
This method is defined in https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/task/ADTaskManager.java#L861-L882 and I moved from there (didn't delete it in this PR, will do it in later PRs). I did it because I felt it would be good to put similar getConfig functions in the same class.

amitgalitz commented on src/main/java/org/opensearch/timeseries/NodeStateManager.java:
```diff
@@ -87,144 +86,67 @@ public NodeStateManager(
      * Used in delete workflow
      *
-     * @param detectorId detector ID
+     * @param configId detector ID
```
nit: remove detector ID here

kaituo replied:
yes, removed

amitgalitz commented on src/main/java/org/opensearch/timeseries/model/Job.java:
```diff
@@ -31,24 +31,22 @@
  * Anomaly detector job.
  */
-public class AnomalyDetectorJob implements Writeable, ToXContentObject, ScheduledJobParameter {
+public class Job implements Writeable, ToXContentObject, ScheduledJobParameter {
```
Has it been decided if we should have an AnomalyDetectorJob and Forecaster Job extending this Job class or use this for both? General should probably be okay to use this for both with some checks for the types

kaituo replied:
I think forecasting and AD share the this class. I will add a type field to job and based on type different [job processors](https://github.com/kaituo/anomaly-detection-1/blob/forecasting11/src/main/java/org/opensearch/timeseries/JobProcessor.java) are running.

amitgalitz commented on src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java:
```diff
@@ -134,12 +136,12 @@ public void runJob(ScheduledJobParameter scheduledJobParameter, JobExecutionCont
-        if (!(scheduledJobParameter instanceof AnomalyDetectorJob)) {
+        if (!(scheduledJobParameter instanceof Job)) {
             throw new IllegalArgumentException(
                 "Job parameter is not instance of AnomalyDetectorJob, type: " + scheduledJobParameter.getClass().getCanonicalName()
```
nit: Message here should probably change to "instance of AnomalyDetectorJob or ForecasterJob" or Just analysis/timeseries job?

kaituo replied:
Changed to "Job parameter is not instance of Job," as forecasting and AD share the same Job class.

amitgalitz commented on src/main/java/org/opensearch/timeseries/NodeState.java:
```diff
@@ -9,198 +9,180 @@
+    // config job
+    private Job configJob;
+
+    // AD only states
```
will this not be needed in forecasting?

kaituo replied:
so far this will not be needed

amitgalitz commented on src/main/java/org/opensearch/timeseries/util/TimeSeriesSafeSecurityInjector.java:
```diff
@@ -48,21 +57,21 @@ public void injectUserRolesFromDetector(ActionListener<Void> injectListener) {
         }
 
-        ActionListener<Optional<AnomalyDetector>> getDetectorListener = ActionListener.wrap(detectorOp -> {
+        ActionListener<Optional<? extends Config>> getDetectorListener = ActionListener.wrap(detectorOp -> {
```
nit: detectorListener can be changed to config or timeseries listener?

kaituo replied:
changed

kaituo commented:
@amitgalitz addressed your comments and also fixing compiler error due to Core's refactor PR https://github.com/opensearch-project/OpenSearch/pull/8636/files
